integrations/PinterestTag/browser.js

```diff
@@ -0,0 +1,220 @@
+          ...pinterestObject,
+        };
+      });
+    });
+    return pinterestObject;
+  }
+
+  /**
+   * This function  simply copies data from rudder payload to new object provided all
+   * the key in properties is present in pinterestPropertySupport
+   * @param {rudder properties} properties
+   * @returns
+   */
+  getRawPayload(properties) {
+    const data = {};
+    Object.keys(properties).forEach((p) => {
+      if (pinterestPropertySupport.includes(p)) {
+        data[p] = properties[p];
+      }
+    });
+    // This logic maps rudder query to search_query for Products Searched events
+    if (isDefinedAndNotNull(properties[searchPropertyMapping.src])) {
+      data[searchPropertyMapping.dest] = properties[searchPropertyMapping.src];
+    }
+    return data;
+  }
+
+  /**
+   * This function will generate required pinterest object to be sent.
+   * getRawPayload() will generate all the destination property excepts lineItems
+   * If rudder payload has products array then line_items is generated
+   * In case if the call is for event which has flag hasEmptyProducts to true, it will generate all
+   * properties including lineItems even if it does not have products array in it ex: Product Added
+   *
+   * @param {rudder payload} properties
+   * @param {*} hasEmptyProducts
+   * @returns
+   */
+  generatePinterestObject(properties, hasEmptyProducts = false) {
+    const pinterestObject = this.getRawPayload(properties);
+
+    let { products } = properties;
+    if (hasEmptyProducts && !products) {
+      products = [properties];
+    }
+    if (products) {
+      const lineItems = [];
+      products.forEach((p) => {
+        const product = this.getMappingObject(p, productPropertyMapping);
+        lineItems.push(product);
+      });
+      pinterestObject.line_items = lineItems;
+    }
+
+    if (this.customProperties.length > 0) {
+      const flattenPayload = flattenJsonPayload(properties);
+      this.customProperties.forEach((custom) => {
+        // This check fails if user is sending boolean value as false
+        // Adding toString because if the property value is boolean then it never gets reflected in destination
+        if (isDefinedAndNotNull(flattenPayload[custom.properties])) {
+          pinterestObject[custom.properties] = flattenPayload[
+            custom.properties
+          ].toString();
+        }
+      });
+    }
+    return pinterestObject;
+  }
+
+  /**
+   * This gives destination events .
+   * Logics: If our eventMapping is not able to map the event that is sent by user payload then it will look into
+   * userDefinedEventsMapping array. In case if it is not found there as well, it will return undefined.
+   * @param {rudder event name} event
+   * @returns
+   */
+  getDestinationEventName(event) {
+    const destinationEvent = eventMapping.find((p) =>
+      p.src.includes(event.toLowerCase())
+    );
+    if (!destinationEvent && this.userDefinedEventsMapping.length > 0) {
```

Comment:
why only check for user defined events mapping when destination events are not present. we should check the user defined mapping anyway and send them.